controllers/virtualmachinegrouppublishrequest/virtualmachinegrouppublishrequest_controller.go

```diff
@@ -95,63 +110,288 @@ func NewReconciler(
+					APIVersion: vmopv1.GroupVersion.String(),
+				},
+			},
+		}
+		if ctx.VMGroupPublishRequest.Spec.Target != "" {
+			request.Spec.Target = vmopv1.VirtualMachinePublishRequestTarget{
+				Location: vmopv1.VirtualMachinePublishRequestTargetLocation{
+					Name:       ctx.VMGroupPublishRequest.Spec.Target,
+					Kind:       reflect.TypeOf(imgregv1a1.ContentLibrary{}).Name(),
+					APIVersion: vmopv1.GroupVersion.String(),
+				},
+			}
+		}
+		if ctx.VMGroupPublishRequest.Spec.TTLSecondsAfterFinished != nil {
+			request.Spec.TTLSecondsAfterFinished = ctx.VMGroupPublishRequest.Spec.TTLSecondsAfterFinished
+		}
+		if err := r.Create(ctx, request); err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return errors.Join(errs...)
+}
+
+func (r *Reconciler) isCompleted(vmGroupPublishReq *vmopv1.VirtualMachineGroupPublishRequest,
+	pendingRequestsSet, completedRequestsSet, desiredRequestsSet sets.Set[string]) bool {
+
+	if completedRequestsSet.Equal(desiredRequestsSet) {
+		conditions.MarkTrue(vmGroupPublishReq, vmopv1.VirtualMachineGroupPublishRequestConditionComplete)
+		vmGroupPublishReq.Status.CompletionTime = metav1.Now()
+		return true
+	}
+
+	missingRequestsSet := desiredRequestsSet.Difference(completedRequestsSet.Union(pendingRequestsSet))
+
+	var reasons, messages []string
+	if len(missingRequestsSet) > 0 {
+		reasons = append(reasons, vmopv1.VirtualMachineGroupPublishRequestConditionReasonMissing)
+		messages = append(messages, fmt.Sprintf("waiting vm publish requests for %s to be created", strings.Join(missingRequestsSet.UnsortedList(), ", ")))
+	}
+
+	if len(pendingRequestsSet) > 0 {
+		reasons = append(reasons, vmopv1.VirtualMachineGroupPublishRequestConditionReasonPending)
+		messages = append(messages, fmt.Sprintf("waiting vm publish requests for %s to be completed", strings.Join(pendingRequestsSet.UnsortedList(), ", ")))
+	}
+
+	conditions.MarkFalse(vmGroupPublishReq,
+		vmopv1.VirtualMachineGroupPublishRequestConditionComplete,
+		strings.Join(reasons, ", "), "%s", strings.Join(messages, ", "))
+
+	return false
+}
+
+func (r *Reconciler) removeAfterTTL(ctx *pkgctx.VirtualMachineGroupPublishRequestContext) error {
+	vmGroupPublishReq := ctx.VMGroupPublishRequest
+
+	if vmGroupPublishReq.Spec.TTLSecondsAfterFinished == nil {
+		return nil
+	}
+
+	expirationTime := vmGroupPublishReq.Status.CompletionTime.Time.Add(time.Duration(*vmGroupPublishReq.Spec.TTLSecondsAfterFinished) * time.Second)
+	if time.Now().Before(expirationTime) {
+		return pkgerr.RequeueError{After: expirationTime.Sub(time.Now())}
+	}
+
+	requests, _, _, err := r.getVMPublishRequests(ctx)
+	if err != nil {
+		return err
+	}
+	if len(requests) != 0 {
+		err := fmt.Errorf("failed to delete vm publish requests after TTL")
+		conditions.MarkError(vmGroupPublishReq,
+			vmopv1.VirtualMachineGroupPublishRequestTTLCondition,
+			fmt.Sprintf("found vm publish requests for %v", slices.Collect(maps.Keys(requests))),
```

Comment:
The order of the keys is going to change on every reconcile causing churn. It isn't clear that we really need to include this in the condition anyways. 

Why does this need to wait for the underlying requests to go away anwyays?